doc/i2p.md

```diff
@@ -47,9 +52,34 @@ In a typical situation, this suffices:
 bitcoind -i2psam=127.0.0.1:7656
 ```
 
-The first time Bitcoin Core connects to the I2P router, its I2P address (and
-corresponding private key) will be automatically generated and saved in a file
-named `i2p_private_key` in the Bitcoin Core data directory.
+The first time Bitcoin Core connects to the I2P router, if a persistent I2P
+address is to be used, then it (and its corresponding private key) will be
+automatically generated and saved in a file named `i2p_private_key` in the
+Bitcoin Core data directory. A persistent I2P address is used for accepting
+incoming connections (`-i2pacceptincoming`) and for making outgoing connections
+if `-i2ptransientout` is not used.
+
+## Persistent vs transient I2P addresses
+
+In I2P connections, the connection receiver sees the I2P address of the
+connection initiator. This is unlike the Tor network where the recipient does
+not know who is connecting to them, not even the initiator's Tor address.
+
+Even if an I2P node is not accepting incoming connections, it is known to other
+nodes by its outgoing I2P address. If this address is always the same, then
+there is an opportunity to whitelist a given node. However, this also creates an
+opportunity to discriminate, fingerprint or analyze a given node based on its
+I2P address.
+
+For this reason, the `-i2ptransientout` configuration option creates a new,
+transient (disposable) I2P address for each new outbound connection. The address
+will never be reused, not even if reconnecting to the same peer later.
+
+`-i2ptransientout` only affects addresses for outgoing connections and not the
+address for accepting incoming connections (if `-i2pacceptincoming` is given).
+The address for accepting connections is advertised to the Bitcoin network and
```

Comment:
Good point! That behavior would be surprising and documenting it is sub-optimal. Better don't advertise our persistent address over I2P outgoing connections that use transient addresses. I extended the commit `net: create a transient I2P session if -i2ptransientout is given` with the following diff to achieve that:

<details>
<summary>[patch] don't self-advertise</summary>

```diff
diff --git a/src/net_processing.cpp b/src/net_processing.cpp
index 74d1bf44d2..17c67ae7a5 100644
--- a/src/net_processing.cpp
+++ b/src/net_processing.cpp
@@ -2773,12 +2773,32 @@ void PeerManagerImpl::ProcessBlock(CNode& node, const std::shared_ptr<const CBlo
     } else {
         LOCK(cs_main);
         mapBlockSource.erase(block->GetHash());
     }
 }
 
+/**
+ * Decide if we should send our own address to a given peer as a self-advertisement.
+ * We do the self-advertisement:
+ * - initially after establishing the connection and
+ * - periodically every ~24h (AVG_LOCAL_ADDRESS_BROADCAST_INTERVAL).
+ */
+static bool ShouldSelfAdvertiseTo(const CNode& node)
+{
+    if (!fListen) {
+        return false;
+    }
+
+    // Don't self-advertise if this is an I2P outbound connection and -i2ptransientout=1.
+    // If we do so, then we would self-advertise our persistent I2P address, used for
+    // accepting incoming connections, which would defeat the purpose of using transient
+    // addresses for outgoing connections.
+    return !node.addrBind.IsI2P() || node.IsInboundConn() ||
+           !gArgs.GetBoolArg("-i2ptransientout", DEFAULT_I2P_TRANSIENT_OUT);
+}
+
 void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type, CDataStream& vRecv,
                                      const std::chrono::microseconds time_received,
                                      const std::atomic<bool>& interruptMsgProc)
 {
     LogPrint(BCLog::NET, "received: %s (%u bytes) peer=%d\n", SanitizeString(msg_type), vRecv.size(), pfrom.GetId());
 
@@ -2924,14 +2944,14 @@ void PeerManagerImpl::ProcessMessage(CNode& pfrom, const std::string& msg_type,
             // empty and no one will know who we are, so these mechanisms are
             // important to help us connect to the network.
             //
             // We skip this for block-relay-only peers. We want to avoid
             // potentially leaking addr information and we do not want to
             // indicate to the peer that we will participate in addr relay.
-            if (fListen && !m_chainman.ActiveChainstate().IsInitialBlockDownload())
-            {
+            if (ShouldSelfAdvertiseTo(pfrom) && !m_chainman.ActiveChainstate().IsInitialBlockDownload()) {
+
                 CAddress addr{GetLocalAddress(pfrom.addr), peer->m_our_services, (uint32_t)GetAdjustedTime()};
                 FastRandomContext insecure_rand;
                 if (addr.IsRoutable())
                 {
                     LogPrint(BCLog::NET, "ProcessMessages: advertising address %s\n", addr.ToString());
                     PushAddress(*peer, addr, insecure_rand);
@@ -4670,13 +4690,13 @@ void PeerManagerImpl::MaybeSendAddr(CNode& node, Peer& peer, std::chrono::micros
 {
     // Nothing to do for non-address-relay peers
     if (!peer.m_addr_relay_enabled) return;
 
     LOCK(peer.m_addr_send_times_mutex);
     // Periodically advertise our local address to the peer.
-    if (fListen && !m_chainman.ActiveChainstate().IsInitialBlockDownload() &&
+    if (ShouldSelfAdvertiseTo(node) && !m_chainman.ActiveChainstate().IsInitialBlockDownload() &&
         peer.m_next_local_addr_send < current_time) {
         // If we've sent before, clear the bloom filter for the peer, so that our
         // self-announcement will actually go out.
         // This might be unnecessary if the bloom filter has already rolled
         // over since our last self-announcement, but there is only a small
         // bandwidth cost that we can incur by doing this (which happens
```
</details>

Since this PR had two ACKs without the above, let me know if it is ok or if I should remove it (@jonatack, @mzumsande).

Reply:
This seems problematic to me: A new node with `-i2pacceptincoming=1` (that doesn't change the default `-i2ptransientout` value) will have a hard time getting inbound peers if it only self-advertises to inbound peers, but only receive inbound peers once it has successfully self-advertised.

Maybe it makes sense to instead soft-set `-i2ptransientout` to false if `-i2pacceptincoming` was chosen? Or, alternatively, just document this behavior, because, as you wrote above, "a node that wants to avoid this should not be listening."

Reply:
A node could self-advertise its I2P address to an IPv4 peer if it only listens on I2P. However, if it also listens on IPv4 then it would prefer its IPv4 address. I guess that for the discussion here we should assume that I2P addresses get self-advertised only (mostly) over I2P connections.

From what you have pointed out it seems that `-i2pacceptincoming=1` is kind of incompatible with `-i2ptransientout=1` because if both are set, then:

* if we self-advertise our persistent address in outgoing I2P connections, then that is very easy to detect and the same as `-i2ptransientout=0` wrt privacy
* if we don't self-advertise in outgoing I2P connections then nobody is going to learn our listening I2P address and we will not get incoming connections, like `-i2pacceptincoming=0`

In current code we have `-i2pacceptincoming` default to `1`. Assuming we don't want to change that, then maybe introduce `-i2ptransientout` with a default of `0` and hard fail the startup if both are explicitly/manually set to `1` by the user? And flip `-i2ptransientout` to `1` if `-i2pacceptincoming` is switched off by the user?

How does this work for Tor? From what I read in the code, if a node is listening on a Tor service, then it would self-advertise its listening Tor address in outbound Tor connections. So, "the Tor peer does not know who is connecting to them" is not true in this case - the peer knows the connecting node's Tor address (which is persistent).

Reply:
> From what you have pointed out it seems that `-i2pacceptincoming=1` is kind of incompatible with `-i2ptransientout=1` 

Yes, I think that's correct.

> In current code we have `-i2pacceptincoming` default to `1`. Assuming we don't want to change that, then maybe introduce `-i2ptransientout` with a default of `0` and hard fail the startup if both are explicitly/manually set to `1` by the user? And flip `-i2ptransientout` to `1` if `-i2pacceptincoming` is switched off by the user?

That looks sensible to me. 

As an alternative thought, I wonder if we would lose much if we'd instead tie the creation of transient addresses to `-i2pacceptincoming=0` and not introduce a `-i2ptransientout` option at all. That would make things much simpler, but we wouldn't support the use case of `-i2pacceptincoming=0`, `-i2ptransientout=0` anymore (so that potential peers couldn't  whitelist our permanent address). But is there a lot utility in whitelisting i2p peers like that? I'm asking because if out peer doesn't know who we are, they don't really have a reason to whitelist us, and if they actually do know us, why would we need to connect to them via i2p in the first place?

> How does this work for Tor? From what I read in the code, if a node is listening on a Tor service, then it would self-advertise its listening Tor address in outbound Tor connections. So, "the Tor peer does not know who is connecting to them" is not true in this case - the peer knows the connecting node's Tor address (which is persistent).

That looks right, although I didn't test it. But I think that if you want to have Tor inbound connections, it is not unreasonable to assume that you might want to let the network (including your outbound peers) know about that.

Reply:
> As an alternative thought...

That sounds even better to me. I like _not_ adding new options :) Use transient addresses if `-i2pacceptincoming=0`, simple. That would make it work similar to Tor - if we are listening then we reveal our (Tor or I2P) address to the peers we connect to (outbound).

As for the downside of not being able to whitelist if `-i2pacceptincoming=0` (and thus transient addresses are being used), that can be done later if https://github.com/bitcoin/bitcoin/pull/17167 gets in, and/or use the persistent address for making outbound connections to nodes given to `-addnode=` or `-connect=` even if transient ones are being used for other outbound connections (due to `-i2pacceptincoming=0`).

Makes sense?

Reply:
> Makes sense?

Yes, I also like that option best! (and the transient address behavior could then be mentioned in the  `-i2pacceptincoming` RPC help).

Reply:
On initial read this SGTM, sensible defaults and not adding the configuration option would also be my preference. Will re-review with this discussion in mind.

Reply:
Done!

Just to answer that one:

> if they actually do know us, why would we need to connect to them via i2p in the first place?

Because we may want to hide the fact that Bitcoin nodes are running on either machine. Also, for the added encryption and authentication based on the I2P address' private key. For example, white-listing IPv4 addresses could be dangerous because, if spoofed, you may end up white-listing a malicious actor who has "taken" that IPv4 address. This is not possible with I2P. To spoof an I2P address one would have to steal the private key from the owner's machine.